### PR TITLE
unsupported encryption config

### PR DESCRIPTION
**Pull request by p0lyn0mial**
allows for specifying an optional prefix for retrieving unsupported config for the encryption controllers

### CODE REVIEW

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go [hunk 1]
```diff
@@ -313,6 +314,7 @@ func (cs *APIServerControllerSet) WithEncryptionControllers(
 	provider controllers.Provider,
 	deployer statemachine.Deployer,
 	migrator migrators.Migrator,
+	rawUnsupportedConfigOverrideFunc config.RawUnsupportedConfigFunc, /*optional can be nil*/
```
Comment (sttts) [inline pkg/operator/apiserver/controllerset/apiservercontrollerset.go:317]: what does it mean if it is nil?

Comment (sttts) [reply pkg/operator/apiserver/controllerset/apiservercontrollerset.go:317]: can we turn this into `WithCustomUnsupportedConfigFunc`?

File: pkg/operator/encryption/config/unsupported_config.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// UnsupportedEncryptionConfig structured config read from the operator's configuration file
+// Usually provided by an external user
+type UnsupportedEncryptionConfig struct {
+	// ForceNewKeyGenerationReason if set forces a new key rotation.
+	ForceNewKeyGenerationReason string
+}
+
+// UnsupportedEncryptionConfigGetter reads an encryption configuration provided by an external user stored in operator's configuration file.
+// customConfigGetter is an optional function provided by an operator for reading its configuration, it takes precedence over getting it from operatorClient.GetOperatorState().
+func UnsupportedEncryptionConfigGetter(operatorClient operatorv1helpers.OperatorClient, customConfigGetter RawUnsupportedConfigFunc) UnsupportedEncryptionConfigFunc {
```
Comment (sttts) [inline pkg/operator/encryption/config/unsupported_config.go:27]: having the client and the custom func is strange here.

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controllerset/apiservercontrollerset.go:317]: if it is nil we will get the config from `operator.spec.UnsupportedConfigOverrides.Raw`

Comment (sttts) [reply pkg/operator/apiserver/controllerset/apiservercontrollerset.go:317]: Wondering whether it isn't much simpler and enough to pass a json path?

File: pkg/operator/encryption/controllers/key_controller.go [hunk 3]
```diff
@@ -336,3 +347,22 @@ func needsNewKey(grKeys state.GroupResourceState, currentMode state.Mode, extern
 	// this also generates back pressure for key rotation when migration takes a long time or was recently completed
 	return latestKeyID, "rotation-interval-has-passed", time.Since(latestKey.Migrated.Timestamp) > encryptionSecretMigrationInterval
 }
+
+// unstructuredUnsupportedConfigFrom returns the configuration from the operator's observedConfig field in the subtree given by the prefix
+func unstructuredUnsupportedConfigFrom(rawConfig []byte, prefix []string) ([]byte, error) {
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/controllers/key_controller.go:352]: deserves a unit test

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go [hunk 4]
```diff
@@ -318,22 +318,27 @@ func (cs *APIServerControllerSet) WithEncryptionControllers(
...
-		migrator,
-		cs.operatorClient,
-		apiServerClient,
-		apiServerInformer,
-		kubeInformersForNamespaces,
-		secretsClient,
-		cs.eventRecorder,
-	)
+
+	cs.encryptionControllers.operatorClient = cs.operatorClient
```
Comment (sttts) [inline pkg/operator/apiserver/controllerset/apiservercontrollerset.go:322]: `cs.encryptionControllers = encryptionControllerBuilder{ .... }`

File: pkg/operator/encryption/controllers/key_controller.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/operator/encryption/controllers/key_controller.go:352]: done.

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go (hunk 4 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controllerset/apiservercontrollerset.go]: done.

**Pull request merged**